File: standalone/lemon-hlwm/php/11-testevents.php
#!/usr/bin/php
<?php # using PHP7

function content_event_idle($pipe_cat_stdin) 
{
    $pid = pcntl_fork();

    switch($pid) {         
    case -1 : // fork errror         
        die('could not fork');
    case 0  : // we are the child
        // start a pipe
        $command_in    = 'herbstclient --idle';
        $pipe_idle_in  = popen($command_in,  'r'); // handle
    
        while(!feof($pipe_idle_in)) {
            # read next event
            $event = fgets($pipe_idle_in);
            fwrite($pipe_cat_stdin, $event);
            flush();
        }
    
        pclose($pipe_idle_in);

        break;
    default : // we are the parent
        // do nothing
        return $pid;
    } 
}

function content_event_interval($pipe_cat_stdin) 
{
    date_default_timezone_set("Asia/Jakarta");
    $pid = pcntl_fork();

    switch($pid) {         
    case -1 : // fork errror         
        die('could not fork');
    case 0  : // we are the child
        $timeformat = '%H:%M:%S';

        do {
            $time_str = strftime($timeformat);
            $time_text = "interval\t${time_str}\n";
            fwrite($pipe_cat_stdin, $time_text);
            flush();
            sleep(3);
        } while (true);
        
        break;
    default : // we are the parent
        // do nothing
        return $pid;
    } 
}

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# main

$descriptorspec = array(
    0 => array('pipe', 'r'),  // stdin
    1 => array('pipe', 'w'),  // stdout
    2 => array('pipe', 'w',)  // stderr
);
    
$proc_cat = proc_open('cat', $descriptorspec, $pipe_cat);

content_event_idle($pipe_cat[0]);
content_event_interval($pipe_cat[0]);

while(!feof($pipe_cat[1])) {
    $event = trim(fgets($pipe_cat[1]));
    echo("$event\n");

    # uncomment to examine more complete behaviour
    # echo("tab:\t[${event}]\n");
}

pclose($pipe_cat[1]);

File: standalone/lemon-hlwm/python/11-testevents.py
#!/usr/bin/env python3

import os
import subprocess

import datetime
import time

def content_event_idle(cat_out):
    pid = os.fork()
    
    if pid == 0:
        try:
            # start a pipe
            command_in = 'herbstclient --idle'  
            pipe_idle_in = subprocess.Popen(
                    [command_in], 
                    stdout = subprocess.PIPE,
                    stderr = subprocess.STDOUT,
                    shell  = True,
                    universal_newlines = True
            )

            # wait for each event  
            for event in pipe_idle_in.stdout: 
                cat_out.stdin.write(event)
                cat_out.stdin.flush()
    
            pipe_idle_in.stdout.close()
        finally:
            import signal
            os.kill(pid, signal.SIGTERM)


def content_event_interval(cat_out):
    pid = os.fork()
    
    if pid == 0:
        try:
            timeformat = '{0:%H:%M:%S}'

            while True:
                time_str  = timeformat.format(datetime.datetime.now())
                time_text = 'interval' + '\t' + time_str + '\n';
    
                cat_out.stdin.write(time_text)
                cat_out.stdin.flush()
    
                time.sleep(3)

            # cat_out.wait()    
        finally:
            import signal
            os.kill(pid, signal.SIGTERM)

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# main

pipe_cat = subprocess.Popen(
        ['cat'], 
        stdin  = subprocess.PIPE,
        stdout = subprocess.PIPE,
        shell  = True,
        universal_newlines=True
    )

content_event_idle(pipe_cat)
content_event_interval(pipe_cat)

for event in pipe_cat.stdout:
    print('event:\t[' + event.strip() +']')

pipe_cat.stdin.close()
pipe_cat.stdout.close()
